test/FunctionalTests/Microsoft.OData.Client.Tests/Tracking/ClientEntityTrackerTests.cs

```diff
@@ -0,0 +1,243 @@
+            ""@odata.context"":""http://localhost:8000/$metadata#ExampleObjectB/$entity"",
+            ""Id"":1,
+            ""Name"":""Example1""}";
+        #endregion
+
+        public ClientEntityTrackerTests()
+        {
+            var uri = new Uri("http://localhost:8000");
+            DefaultTrackingContext = new Container(uri);
+        }
+
+        [Fact]
+        public async Task SavingObject_ThatOveridesEqualsAndGetHashCodeMethods_SavesSuccessfully()
+        {
+            SetupContextWithRequestPipelineForSaving(
+             new DataServiceContext[] { DefaultTrackingContext }, "ExampleObject");
+            var exampleObject = new ExampleObject
+            {
+                Id = 0,
+                Name = "Example1"
+            };
+
+            DefaultTrackingContext.AddObject("ExampleObjects", exampleObject);
+            Assert.NotNull(DefaultTrackingContext.GetEntityDescriptor(exampleObject));
+            await SaveContextChangesAsync(new DataServiceContext[] { DefaultTrackingContext });
+
+        }
+
+        [Fact]
+        public async Task SavingObject_ThatImplementsIEqualityComparerInterface_SavesSuccessfully()
+        {
+            SetupContextWithRequestPipelineForSaving(
+             new DataServiceContext[] { DefaultTrackingContext }, "ExampleObjectA");
+            var exampleObject = new ExampleObjectA
+            {
+                Id = 0,
+                Name = "Example1"
+            };
+
+            DefaultTrackingContext.AddObject("ExampleObjectAs", exampleObject);
+            Assert.NotNull(DefaultTrackingContext.GetEntityDescriptor(exampleObject));
+            await SaveContextChangesAsync(new DataServiceContext[] { DefaultTrackingContext });
+
+        }
+
+        [Fact]
+        public async Task SavingObject_ThatDoesNotImplementOrOvverideEqualityComparer_SavesSuccessfully()
+        {
+            SetupContextWithRequestPipelineForSaving(
+             new DataServiceContext[] { DefaultTrackingContext }, "ExampleObjectB");
+            var exampleObject = new ExampleObjectB
+            {
+                Id = 0,
+                Name = "Example1"
+            };
+
+            DefaultTrackingContext.AddObject("ExampleObjectAs", exampleObject);
+            Assert.NotNull(DefaultTrackingContext.GetEntityDescriptor(exampleObject));
+            await SaveContextChangesAsync(new DataServiceContext[] { DefaultTrackingContext });
+
+        }
+
+        private async Task SaveContextChangesAsync(DataServiceContext[] dataServiceContexts)
+        {
+            foreach (var dataServiceContext in dataServiceContexts)
+            {
+                await dataServiceContext.SaveChangesAsync();
+            }
+        }
+
+        private void SetupContextWithRequestPipelineForSaving(DataServiceContext[] dataServiceContexts, string type)
+        {
+            var response = "";
+            var location = "";
+            if (type == "ExampleObject")

```

Comment:
Refactor this to move the type out of the method and pass the expected response and location so as to reduce the length of the method.